File: qt/src/drivemusicapi.cpp
#include "drivemusicapi.h"

#include <QDir>
#include <QFile>
#include <QRegularExpression>
#include <QSet>
#include <QUrl>
#include <algorithm>

#include "appconstants.h"
#include "httpclient.h"

using namespace AppConstants;

static QString decodeHtml(const QString &s) {
    QString r = s;
    r.replace(QStringLiteral("&amp;"), QStringLiteral("&"));
    r.replace(QStringLiteral("&quot;"), QStringLiteral("\""));
    r.replace(QStringLiteral("&#39;"), QStringLiteral("'"));
    return r;
}

static QString pageUrl(const Track &track) {
    const QString u = track.url.trimmed();
    if (u.contains(QStringLiteral("drivemusic.me")) && u.endsWith(QStringLiteral(".html"))) {
        return u;
    }
    if (u.startsWith(QLatin1Char('/')) && u.endsWith(QStringLiteral(".html"))) {
        return kDriveMusicBase + u;
    }
    return {};
}

static QStringList extractMp3Urls(const QString &html) {
    QStringList urls;
    QRegularExpression re(
        QStringLiteral("https://[a-z0-9.-]*drivemusic\\.me/dl/[^\"\\s<>]+\\.mp3"),
        QRegularExpression::CaseInsensitiveOption);
    auto it = re.globalMatch(html);
    while (it.hasNext()) {
        urls.append(it.next().captured(0));
    }
    urls.removeDuplicates();
    std::sort(urls.begin(), urls.end(), [](const QString &a, const QString &b) {
        const bool ao = a.contains(QStringLiteral("/dl/online/"));
        const bool bo = b.contains(QStringLiteral("/dl/online/"));
        if (ao != bo) {
            return ao;
        }
        return a.length() < b.length();
    });
    return urls;
}

QVector<Track> DriveMusicApi::search(const QString &query, QString *error) {
    const QUrl url(kDriveMusicBase + QStringLiteral("/?do=search&subaction=search&story=")
                   + QString::fromUtf8(QUrl::toPercentEncoding(query.trimmed())));

    QMap<QString, QString> headers;
    headers.insert(QStringLiteral("Referer"), kDriveMusicBase);

    const auto resp = HttpClient::get(url, headers);
    if (!resp.ok()) {
        if (error) {
            *error = resp.error.isEmpty()
                         ? QStringLiteral("HTTP %1").arg(resp.status)
                         : resp.error;
        }
        return {};
    }

    const QString body = QString::fromUtf8(resp.body);
    QVector<Track> results;
    QSet<QString> seen;

    QRegularExpression itemRe(
        QStringLiteral(
            "(?s)href=\"(/[a-z0-9_]+/(\\d+)-[^\"]+\\.html)\"[^>]*class=\"popular-play-author\"[^>]*>([^<]*)</a>"
            ".*?popular-play-composition.*?(?:<a[^>]*>)?([^<]*)"),
        QRegularExpression::CaseInsensitiveOption);

    auto it = itemRe.globalMatch(body);
    while (it.hasNext()) {
        const auto m = it.next();
        const QString id = m.captured(2).trimmed();
        if (id.isEmpty() || seen.contains(id)) {
            continue;
        }
        seen.insert(id);
        const QString path = m.captured(1).trimmed();
        const QString page = path.startsWith(QStringLiteral("http")) ? path : kDriveMusicBase + path;
        Track t;
        t.id = id;
        t.title = decodeHtml(m.captured(3).trimmed());
        t.artist = decodeHtml(m.captured(4).trimmed());
        t.url = page;
        t.source = DownloadSource::DriveMusic;
        if (!t.title.isEmpty()) {
            results.append(t);
        }
    }

    if (results.isEmpty() && error) {
        *error = QStringLiteral("Ничего не найдено на DriveMusic по запросу «%1».").arg(query);
    }
    return results.mid(0, 30);
}

static QString sanitizeName(const QString &s) {
    QString r = s.trimmed();
    const QString bad = QStringLiteral("/\\:*?\"<>|");
    for (const QChar c : bad) {
        r.replace(c, QLatin1Char('_'));
    }
    return r;
}

QString DriveMusicApi::download(const Track &track,
                                const QString &folder,
                                QString *error) {
    const QString page = pageUrl(track);
    if (page.isEmpty()) {
        if (error) {
            *error = QStringLiteral(
                "DriveMusic: нет ссылки на страницу трека — найдите трек через поиск.");
        }
        return {};
    }

    QMap<QString, QString> headers;
    headers.insert(QStringLiteral("Referer"), kDriveMusicBase);

    const auto pageResp = HttpClient::get(QUrl(page), headers);
    if (!pageResp.ok()) {
        if (error) {
            *error = QStringLiteral("Страница: %1").arg(pageResp.error);
        }
        return {};
    }

    const auto urls = extractMp3Urls(QString::fromUtf8(pageResp.body));
    const QString dest = QDir(folder).filePath(
        sanitizeName(track.artist + QStringLiteral(" - ") + track.title + QStringLiteral("_")
                      + track.id)
        + QStringLiteral(".mp3"));

    QMap<QString, QString> dlHeaders = headers;
    dlHeaders.insert(QStringLiteral("Referer"), page);
    dlHeaders.insert(QStringLiteral("Origin"), kDriveMusicBase);
    dlHeaders.insert(QStringLiteral("Accept"),
                     QStringLiteral("audio/mpeg,application/octet-stream,*/*;q=0.8"));

    QString lastErr;
    for (const QString &url : urls) {
        const auto r = HttpClient::downloadToFile(QUrl(url), dest, dlHeaders);
        if (r.ok()) {
            QFile f(dest);
            if (f.exists() && f.size() >= kMinDownloadBytes) {
                return dest;
            }
            f.remove();
            lastErr = QStringLiteral("файл слишком маленький");
        } else {
            lastErr = r.error;
        }
        QFile::remove(dest);
    }

    if (error) {
        *error = QStringLiteral("DriveMusic: %1").arg(lastErr);
    }
    return {};
}

QString DriveMusicApi::streamUrl(const Track &track, QString *error) {
    const QString page = pageUrl(track);
    QMap<QString, QString> headers;
    headers.insert(QStringLiteral("Referer"), kDriveMusicBase);
    const auto pageResp = HttpClient::get(QUrl(page), headers);
    if (!pageResp.ok()) {
        if (error) {
            *error = pageResp.error;
        }
        return {};
    }
    const auto urls = extractMp3Urls(QString::fromUtf8(pageResp.body));
    QString u;
    for (const QString &url : urls) {
        if (url.contains(QStringLiteral("/dl/online/"))) {
            u = url;
            break;
        }
    }
    if (u.isEmpty() && !urls.isEmpty()) {
        u = urls.first();
    }
    if (u.isEmpty() && error) {
        *error = QStringLiteral("На странице нет MP3 URL");
    }
    return u;
}

File: android/app/src/main/java/net/mp3party/downloader/DriveMusicApi.kt
package net.mp3party.downloader

import okhttp3.Cookie
import okhttp3.CookieJar
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.net.URLEncoder
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern
import java.net.URLDecoder

object DriveMusicApi {
    private const val BASE = "https://ru.drivemusic.me"
    private const val USER_AGENT =
        "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Mobile Safari/537.36"
    private const val MIN_DOWNLOAD_BYTES = 50 * 1024

    private val cookieStore = ConcurrentHashMap<String, MutableList<Cookie>>()

    private val client = OkHttpClient.Builder()
        .cookieJar(object : CookieJar {
            override fun saveFromResponse(url: HttpUrl, cookies: List<Cookie>) {
                cookieStore.getOrPut(url.host) { mutableListOf() }.apply {
                    cookies.forEach { c ->
                        removeAll { it.name == c.name }
                        add(c)
                    }
                }
            }

            override fun loadForRequest(url: HttpUrl): List<Cookie> =
                cookieStore[url.host].orEmpty()
        })
        .followRedirects(true)
        .followSslRedirects(true)
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .writeTimeout(120, TimeUnit.SECONDS)
        .build()

    private val mp3UrlPattern = Pattern.compile(
        """https://[a-z0-9.-]*drivemusic\.me/dl/[^"\s<>]+\.mp3""",
        Pattern.CASE_INSENSITIVE,
    )

    private val searchItemPattern = Pattern.compile(
        """(?s)href="(/[a-z0-9_]+/(\d+)-[^"]+\.html)"[^>]*class="popular-play-author"[^>]*>([^<]*)</a>.*?popular-play-composition.*?>(?:<a[^>]*>)?([^<]*)""",
        Pattern.CASE_INSENSITIVE,
    )

    fun search(query: String): List<Track> {
        val encoded = URLEncoder.encode(query.trim(), Charsets.UTF_8.name())
        val url = "$BASE/?do=search&subaction=search&story=$encoded"
        val body = get(url, BASE) ?: return emptyList()

        val results = mutableListOf<Track>()
        val seen = mutableSetOf<String>()
        val m = searchItemPattern.matcher(body)
        while (m.find()) {
            val id = m.group(2)?.trim().orEmpty()
            if (id.isEmpty() || !seen.add(id)) continue
            val title = decode(m.group(3)?.trim().orEmpty())
            val artist = decode(m.group(4)?.trim().orEmpty())
            if (title.isEmpty()) continue
            val path = m.group(1)?.trim().orEmpty()
            val pageUrl = if (path.startsWith("http")) path else "$BASE$path"
            results.add(
                Track(
                    id = id,
                    artist = artist,
                    title = title,
                    streamUrl = pageUrl,
                    source = DownloadSource.DriveMusic,
                ),
            )
        }
        return results.take(30)
    }

    fun download(track: Track, destFile: File): String {
        val pageUrl = pageUrl(track)
        val pageBody = get(pageUrl, BASE) ?: ""
        val candidates = downloadCandidates(pageBody, track)

        var lastErr = "нет доступных URL"
        for (url in candidates) {
            try {
                val request = Request.Builder()
                    .url(url)
                    .header("User-Agent", USER_AGENT)
                    .header("Referer", pageUrl)
                    .header("Origin", BASE)
                    .header("Accept", "audio/mpeg,application/octet-stream,*/*;q=0.8")
                    .header("Accept-Language", "ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7")
                    .build()

                client.newCall(request).execute().use { resp ->
                    if (!resp.isSuccessful) {
                        lastErr = "HTTP ${resp.code}"
                        return@use
                    }
                    val body = resp.body ?: run {
                        lastErr = "пустой ответ"
                        return@use
                    }
                    destFile.parentFile?.mkdirs()
                    var written = 0L
                    destFile.outputStream().use { out ->
                        body.byteStream().use { input ->
                            val buf = ByteArray(8192)
                            while (true) {
                                val n = input.read(buf)
                                if (n <= 0) break
                                out.write(buf, 0, n)
                                written += n
                            }
                        }
                    }
                    if (written < MIN_DOWNLOAD_BYTES) {
                        destFile.delete()
                        lastErr = "файл слишком маленький ($written B), ссылка могла устареть"
                        return@use
                    }

                    // Переименовать файл по Content-Disposition
                    val cdHeader = resp.header("Content-Disposition") ?: ""
                    val cdName = extractFileNameFromDisposition(cdHeader)
                    if (cdName != null && cdName.endsWith(".mp3", ignoreCase = true)) {
                        val cleaned = MusicLibrary.cleanDispositionFilename(cdName)
                        val renamed = File(destFile.parentFile, cleaned)
                        if (renamed.absolutePath != destFile.absolutePath) {
                            destFile.renameTo(renamed)
                            return renamed.absolutePath
                        }
                    }

                    return destFile.absolutePath
                }
            } catch (e: Exception) {
                lastErr = e.message ?: e.toString()
                destFile.delete()
            }
        }
        throw IllegalStateException("DriveMusic: $lastErr")
    }

    /** Прямой URL для стрима (online), если есть на странице. */
    fun resolveStreamUrl(track: Track): String? {
        val pageUrl = pageUrl(track)
        val body = get(pageUrl, BASE) ?: return null
        val urls = extractMp3Urls(body)
        return urls.firstOrNull { it.contains("/dl/online/") }
            ?: urls.firstOrNull()
    }

    private fun pageUrl(track: Track): String {
        val u = track.streamUrl.trim()
        when {
            u.contains("drivemusic.me") && u.endsWith(".html") -> return normalizeUrl(u)
            u.startsWith("/") && u.endsWith(".html") -> return "$BASE$u"
        }
        throw IllegalStateException(
            "DriveMusic: нет ссылки на страницу трека — найдите трек через поиск.",
        )
    }

    private fun downloadCandidates(pageBody: String, track: Track): List<String> {
        val urls = extractMp3Urls(pageBody).toMutableList()
        val sorted = urls.distinct().sortedWith(
            compareBy(
                { if (it.contains("/dl/online/")) 1 else 0 },
                { it.length },
            ),
        )
        return sorted
    }

    private fun extractMp3Urls(html: String): List<String> {
        val out = mutableListOf<String>()
        val m = mp3UrlPattern.matcher(html)
        while (m.find()) {
            out.add(m.group())
        }
        return out
    }

    private fun normalizeUrl(url: String): String {
        var u = decode(url.trim())
        if (u.startsWith("//")) u = "https:$u"
        if (u.startsWith("/")) u = "$BASE$u"
        return u
    }

    private fun extractFileNameFromDisposition(header: String): String? {
        val utf8 = Regex("filename\\*=UTF-8''([^;\\s]+)", RegexOption.IGNORE_CASE)
            .find(header)?.groupValues?.get(1)
        if (utf8 != null) {
            return try { URLDecoder.decode(utf8, "UTF-8") } catch (_: Exception) { null }
        }
        val plain = Regex("""filename="?([^";\s]+)"?""", RegexOption.IGNORE_CASE)
            .find(header)?.groupValues?.get(1)
        if (plain != null) {
            return try { URLDecoder.decode(plain, "UTF-8") } catch (_: Exception) { null }
        }
        return null
    }

    private fun get(url: String, referer: String?): String? {
        val builder = Request.Builder()
            .url(url)
            .header("User-Agent", USER_AGENT)
            .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
            .header("Accept-Language", "ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7")
        if (referer != null) builder.header("Referer", referer)
        client.newCall(builder.build()).execute().use { resp ->
            if (!resp.isSuccessful) return null
            return resp.body?.string()
        }
    }

    private fun decode(s: String): String =
        s.replace("&amp;", "&")
            .replace("&quot;", "\"")
            .replace("&#39;", "'")
            .replace("&lt;", "<")
            .replace("&gt;", ">")
}
